src/jit/codegenarmarch.cpp

```diff
@@ -1922,228 +1922,259 @@ void CodeGen::genCodeForCpBlkHelper(GenTreeBlk* cpBlkNode)
-    }
+    regNumber dstAddrBaseReg = genConsumeReg(node->Addr());
+    unsigned  dstOffset      = 0;
 
-    assert(dstAddr->isUsedFromReg());
-    assert((initVal->isUsedFromReg() && !initVal->IsIntegralConst(0)) || initVal->IsIntegralConst(0));
-    assert(size != 0);
-    assert(size <= INITBLK_UNROLL_LIMIT);
+    regNumber srcReg;
+    GenTree*  src = node->Data();
 
-    emitter* emit = GetEmitter();
+    if (src->OperIs(GT_INIT_VAL))
+    {
+        assert(src->isContained());
+        src = src->gtGetOp1();
+    }
 
-    genConsumeOperands(initBlkNode);
+    if (!src->isContained())
+    {
+        srcReg = genConsumeReg(src);
+    }
+    else
+    {
+        assert(src->IsIntegralConst(0));
+        srcReg = REG_ZR;
+    }
 
-    if (initBlkNode->gtFlags & GTF_BLK_VOLATILE)
+    if (node->IsVolatile())
     {
-        // issue a full memory barrier before a volatile initBlockUnroll operation
         instGen_MemoryBarrier();
     }
 
-    regNumber valReg = initVal->IsIntegralConst(0) ? REG_ZR : initVal->gtRegNum;
-
-    assert(!initVal->IsIntegralConst(0) || (valReg == REG_ZR));
-
-    unsigned offset = 0;
+    emitter* emit = GetEmitter();
+    unsigned size = node->GetLayout()->GetSize();
 
-    // Perform an unroll using stp.
-    if (size >= 2 * REGSIZE_BYTES)
+    for (unsigned regSize = 2 * REGSIZE_BYTES; size >= regSize; size -= regSize, dstOffset += regSize)
     {
-        // Determine how many 16 byte slots
-        size_t slots = size / (2 * REGSIZE_BYTES);
-
-        while (slots-- > 0)
-        {
-            emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, valReg, valReg, dstAddr->gtRegNum, offset);
-            offset += (2 * REGSIZE_BYTES);
-        }
+        emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, srcReg, srcReg, dstAddrBaseReg, dstOffset);
     }
 
-    // Fill the remainder (15 bytes or less) if there's any.
-    if ((size & 0xf) != 0)
+    for (unsigned regSize = REGSIZE_BYTES; size > 0; size -= regSize, dstOffset += regSize)
     {
-        if ((size & 8) != 0)
-        {
-            emit->emitIns_R_R_I(INS_str, EA_8BYTE, valReg, dstAddr->gtRegNum, offset);
-            offset += 8;
-        }
-        if ((size & 4) != 0)
+        while (regSize > size)
         {
-            emit->emitIns_R_R_I(INS_str, EA_4BYTE, valReg, dstAddr->gtRegNum, offset);
-            offset += 4;
+            regSize /= 2;
         }
-        if ((size & 2) != 0)
-        {
-            emit->emitIns_R_R_I(INS_strh, EA_2BYTE, valReg, dstAddr->gtRegNum, offset);
-            offset += 2;
-        }
-        if ((size & 1) != 0)
+
+        instruction storeIns;
+
+        switch (regSize)
         {
-            emit->emitIns_R_R_I(INS_strb, EA_1BYTE, valReg, dstAddr->gtRegNum, offset);
+            case 1:
+                storeIns = INS_strb;
+                break;
+            case 2:
+                storeIns = INS_strh;
+                break;
+            default:
+                storeIns = INS_str;
```

Comment:
Yes, I suppose the loop may look a bit strange. But it's very convenient to avoid duplication and keep the whole thing in one place. I'll update the switch.